// @logisticinfotech/ionic4-datepicker/esm2015/lib/ionic4-datepicker.module.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/ionic4-datepicker.module.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { FormsModule } from '@angular/forms';
import { LiIonic4DatepickerDirective } from './li-ionic4-datepicker.directive';
import { Ionic4DatepickerModalComponent } from './ionic4-datepicker-modal/ionic4-datepicker-modal.component';
import { IonicModule } from '@ionic/angular';
import { Ionic4DatepickerComponent } from './ionic4-datepicker.component';
export class Ionic4DatepickerModule {
}
Ionic4DatepickerModule.decorators = [
    { type: NgModule, args: [{
                imports: [
                    CommonModule,
                    FormsModule,
                    IonicModule
                ],
                declarations: [
                    Ionic4DatepickerComponent,
                    LiIonic4DatepickerDirective,
                    Ionic4DatepickerModalComponent
                ],
                exports: [
                    Ionic4DatepickerComponent,
                    LiIonic4DatepickerDirective,
                    Ionic4DatepickerModalComponent,
                    CommonModule,
                    FormsModule
                ],
                entryComponents: [
                    // Ionic4DatepickerComponent,
                    Ionic4DatepickerModalComponent
                ],
                providers: []
            },] }
];
