// ngx-angular-autocomplete/__ivy_ngcc__/fesm2015/ngx-angular-autocomplete.js
import { Injectable, ɵɵdefineInjectable, EventEmitter, Component, Input, ViewChild, Output, Directive, ElementRef, HostListener, NgModule } from '@angular/core';
import { of, fromEvent } from 'rxjs';
import { map, debounceTime, tap, switchMap } from 'rxjs/operators';
import { CommonModule } from '@angular/common';

/**
 * @fileoverview added by tsickle
 * Generated from: lib/ngx-autocomplete.service.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/common';

const _c0 = ["searchInput"];
const _c1 = function (a0) { return { "suggestion-active": a0 }; };
function NgxAutocompleteComponent_a_4_Template(rf, ctx) { if (rf & 1) {
    const _r7 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "a", 6);
    ɵngcc0.ɵɵlistener("click", function NgxAutocompleteComponent_a_4_Template_a_click_0_listener() { ɵngcc0.ɵɵrestoreView(_r7); const entry_r4 = ctx.$implicit; const ctx_r6 = ɵngcc0.ɵɵnextContext(); return ctx_r6.setValue(entry_r4); });
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const entry_r4 = ctx.$implicit;
    const i_r5 = ctx.index;
    const ctx_r1 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵpropertyInterpolate1("id", "list", i_r5, "");
    ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction1(3, _c1, i_r5 == ctx_r1.activeSuggestion));
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate1(" ", entry_r4[ctx_r1.searchKeyword], " ");
} }
function NgxAutocompleteComponent_a_5_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "a");
    ɵngcc0.ɵɵtext(1, "No Data Found");
    ɵngcc0.ɵɵelementEnd();
} }
function NgxAutocompleteComponent_a_6_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "a");
    ɵngcc0.ɵɵtext(1, "Searching...");
    ɵngcc0.ɵɵelementEnd();
} }
class NgxAutocompleteService {
    constructor() { }
}
NgxAutocompleteService.ɵfac = function NgxAutocompleteService_Factory(t) { return new (t || NgxAutocompleteService)(); };
NgxAutocompleteService.ɵprov = ɵngcc0.ɵɵdefineInjectable({ token: NgxAutocompleteService, factory: NgxAutocompleteService.ɵfac, providedIn: 'root' });
/** @nocollapse */
NgxAutocompleteService.ctorParameters = () => [];
/** @nocollapse */ NgxAutocompleteService.ngInjectableDef = ɵɵdefineInjectable({ factory: function NgxAutocompleteService_Factory() { return new NgxAutocompleteService(); }, token: NgxAutocompleteService, providedIn: "root" });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(NgxAutocompleteService, [{
        type: Injectable,
        args: [{
                providedIn: 'root'
            }]
    }], function () { return []; }, null); })();

/**
 * @fileoverview added by tsickle
 * Generated from: lib/ngx-autocomplete.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
class NgxAutocompleteComponent {
    constructor() {
        this.selectedValue = new EventEmitter();
        this.data = [];
        this.showSearches = false;
        this.isSearching = false;
        this.activeSuggestion = 0;
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        this.dataSearch();
    }
    /**
     * @param {?} name
     * @return {?}
     */
    getData(name) {
        /** @type {?} */
        let data = this.entries.filter((/**
         * @param {?} val
         * @return {?}
         */
        (val) => {
            return val[this.searchKeyword].toLowerCase().includes(name.toLowerCase()) == true;
        }));
        return of(data);
    }
    /**
     * @return {?}
     */
    dataSearch() {
        /** @type {?} */
        const search$ = fromEvent(this.searchInput.nativeElement, 'input').pipe(map((/**
         * @param {?} event
         * @return {?}
         */
        (event) => event.target.value)), debounceTime(500), tap((/**
         * @return {?}
         */
        () => this.isSearching = true)), switchMap((/**
         * @param {?} term
         * @return {?}
         */
        (term) => term ? this.getData(term) : of([]))), tap((/**
         * @return {?}
         */
        () => {
            this.isSearching = false,
                this.showSearches = true;
        })));
        search$.subscribe((/**
         * @param {?} data
         * @return {?}
         */
        data => {
            this.isSearching = false;
            this.data = data;
        }));
    }
    /**
     * @param {?} data
     * @return {?}
     */
    setValue(data) {
        /** @type {?} */
        let value = data[this.searchKeyword];
        this.searchInput.nativeElement.value = value;
        this.data = [data];
        this.activeSuggestion = 0;
        this.selectedValue.emit(data);
        this.showSearches = false;
    }
    /**
     * @param {?} index
     * @param {?} item
     * @return {?}
     */
    trackById(index, item) {
        return item._id;
    }
    /**
     * @return {?}
     */
    closeDropDown() {
        this.showSearches = false;
    }
    /**
     * @param {?} e
     * @return {?}
     */
    changeSuggestion(e) {
        if (e.keyCode === 13) {
            if (this.data.length != 0) {
                this.setValue(this.data[this.activeSuggestion]);
                this.scrollTo();
            }
        }
        // User pressed the up arrow
        else if (e.keyCode == 38) {
            if (this.activeSuggestion == 0) {
                return false;
            }
            else {
                this.activeSuggestion = this.activeSuggestion - 1;
                this.scrollTo();
            }
        }
        // User pressed the down arrow
        else if (e.keyCode == 40) {
            if (this.activeSuggestion == this.data.length - 1) {
                return false;
            }
            else {
                this.activeSuggestion = this.activeSuggestion + 1;
                this.scrollTo();
            }
        }
    }
    /**
     * @private
     * @return {?}
     */
    scrollTo() {
        /** @type {?} */
        let str = `list${this.activeSuggestion}`;
        /** @type {?} */
        let elmnt = document.getElementById(str);
        elmnt.scrollIntoView();
        window.scrollTo(0, 0);
    }
}
NgxAutocompleteComponent.ɵfac = function NgxAutocompleteComponent_Factory(t) { return new (t || NgxAutocompleteComponent)(); };
NgxAutocompleteComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: NgxAutocompleteComponent, selectors: [["ngx-autocomplete"]], viewQuery: function NgxAutocompleteComponent_Query(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵstaticViewQuery(_c0, true);
    } if (rf & 2) {
        var _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.searchInput = _t.first);
    } }, inputs: { entries: "entries", placeholderValue: "placeholderValue", inputId: "inputId", searchKeyword: "searchKeyword" }, outputs: { selectedValue: "selectedValue" }, decls: 7, vars: 9, consts: [[1, "position-relative"], ["type", "text", "id", "auto-search", "autocomplete", "off", 1, "form-control", 3, "placeholder", "id", "click", "keyup"], ["searchInput", ""], ["clickOutside", "", 1, "dropdown-content", 3, "id", "inputId", "hidden", "clickOutside"], ["tappable", "", 3, "ngClass", "id", "click", 4, "ngFor", "ngForOf", "ngForTrackBy"], [4, "ngIf"], ["tappable", "", 3, "ngClass", "id", "click"]], template: function NgxAutocompleteComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "div", 0);
        ɵngcc0.ɵɵelementStart(1, "input", 1, 2);
        ɵngcc0.ɵɵlistener("click", function NgxAutocompleteComponent_Template_input_click_1_listener() { return ctx.showSearches = true; })("keyup", function NgxAutocompleteComponent_Template_input_keyup_1_listener($event) { return ctx.changeSuggestion($event); });
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(3, "div", 3);
        ɵngcc0.ɵɵlistener("clickOutside", function NgxAutocompleteComponent_Template_div_clickOutside_3_listener() { return ctx.closeDropDown(); });
        ɵngcc0.ɵɵtemplate(4, NgxAutocompleteComponent_a_4_Template, 2, 5, "a", 4);
        ɵngcc0.ɵɵtemplate(5, NgxAutocompleteComponent_a_5_Template, 2, 0, "a", 5);
        ɵngcc0.ɵɵtemplate(6, NgxAutocompleteComponent_a_6_Template, 2, 0, "a", 5);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("placeholder", ctx.placeholderValue)("id", ctx.inputId + 1);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("id", ctx.inputId + 2)("inputId", ctx.inputId + 1)("hidden", !ctx.showSearches);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngForOf", ctx.data)("ngForTrackBy", ctx.trackById);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.data && ctx.data.length <= 0 && !ctx.isSearching);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.isSearching);
    } }, directives: function () { return [DropdownDirective, ɵngcc1.NgForOf, ɵngcc1.NgIf, ɵngcc1.NgClass]; }, styles: [".dropdown-content[_ngcontent-%COMP%] {\n      background-color: #fafafa;\n      width: 100%;\n      max-height: 150px;\n      overflow-y: auto;\n      box-shadow: 0px 8px 16px 0px rgba(0,0,0,0.2);\n      z-index: 1;\n      cursor: pointer;\n    }\n\n    .dropdown-content[_ngcontent-%COMP%]   a[_ngcontent-%COMP%] {\n      color: black;\n      padding: 15px 18px;\n      text-decoration: none;\n      display: block;\n    }\n\n    .suggestion-active[_ngcontent-%COMP%], .dropdown-content[_ngcontent-%COMP%]   a[_ngcontent-%COMP%]:hover {background-color: #f1f1f1}\n\n    input[_ngcontent-%COMP%] {\n      width: 100%;\n      padding: 12px 20px;\n      margin: 8px 0;\n      box-sizing: border-box;\n    }"] });
/** @nocollapse */
NgxAutocompleteComponent.ctorParameters = () => [];
NgxAutocompleteComponent.propDecorators = {
    entries: [{ type: Input }],
    placeholderValue: [{ type: Input }],
    inputId: [{ type: Input }],
    searchKeyword: [{ type: Input }],
    searchInput: [{ type: ViewChild, args: ['searchInput', { static: true },] }],
    selectedValue: [{ type: Output }]
};
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(NgxAutocompleteComponent, [{
        type: Component,
        args: [{
                selector: 'ngx-autocomplete',
                template: `
    <div class="position-relative">

    <input type="text" class="form-control" [placeholder]="placeholderValue" id="auto-search" #searchInput autocomplete="off" [id]="inputId+1" (click)="showSearches = true" (keyup)="changeSuggestion($event)">
    
    <div [id]="inputId+2" clickOutside (clickOutside)="closeDropDown()" [inputId]="inputId+1" [hidden]="!showSearches" class="dropdown-content">
        <a *ngFor="let entry of data; trackBy: trackById;let i = index" tappable (click)="setValue(entry)"  [ngClass]="{ 'suggestion-active' : i == activeSuggestion} " id="list{{i}}">
                {{entry[searchKeyword] }}   
        </a>
        <a *ngIf="data && data.length <= 0 && !isSearching" >No Data Found</a>
        <a *ngIf="isSearching">Searching...</a>
    </div>
    
  </div>
  `,
                styles: [`
  
    .dropdown-content {
      background-color: #fafafa;
      width: 100%;
      max-height: 150px;
      overflow-y: auto;
      box-shadow: 0px 8px 16px 0px rgba(0,0,0,0.2);
      z-index: 1;
      cursor: pointer;
    }

    .dropdown-content a {
      color: black;
      padding: 15px 18px;
      text-decoration: none;
      display: block;
    }

    .suggestion-active, .dropdown-content a:hover {background-color: #f1f1f1}

    input {
      width: 100%;
      padding: 12px 20px;
      margin: 8px 0;
      box-sizing: border-box;
    }
 
  `]
            }]
    }], function () { return []; }, { selectedValue: [{
            type: Output
        }], entries: [{
            type: Input
        }], placeholderValue: [{
            type: Input
        }], inputId: [{
            type: Input
        }], searchKeyword: [{
            type: Input
        }], searchInput: [{
            type: ViewChild,
            args: ['searchInput', { static: true }]
        }] }); })();
if (false) {
    /** @type {?} */
    NgxAutocompleteComponent.prototype.entries;
    /** @type {?} */
    NgxAutocompleteComponent.prototype.placeholderValue;
    /** @type {?} */
    NgxAutocompleteComponent.prototype.inputId;
    /** @type {?} */
    NgxAutocompleteComponent.prototype.searchKeyword;
    /** @type {?} */
    NgxAutocompleteComponent.prototype.searchInput;
    /** @type {?} */
    NgxAutocompleteComponent.prototype.selectedValue;
    /** @type {?} */
    NgxAutocompleteComponent.prototype.data;
    /** @type {?} */
    NgxAutocompleteComponent.prototype.showSearches;
    /** @type {?} */
    NgxAutocompleteComponent.prototype.isSearching;
    /** @type {?} */
    NgxAutocompleteComponent.prototype.activeSuggestion;
    /** @type {?} */
    NgxAutocompleteComponent.prototype.searchValue;
}

/**
 * @fileoverview added by tsickle
 * Generated from: lib/dropdown.directive.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
class DropdownDirective {
    /**
     * @param {?} _elementRef
     */
    constructor(_elementRef) {
        this._elementRef = _elementRef;
        this.clickOutside = new EventEmitter();
    }
    /**
     * @param {?} targetElement
     * @return {?}
     */
    onClick(targetElement) {
        if (targetElement.id == this.inputId) {
            return;
        }
        /** @type {?} */
        const isClickedInside = this._elementRef.nativeElement.contains(targetElement);
        if (!isClickedInside) {
            this.clickOutside.emit(null);
        }
    }
}
DropdownDirective.ɵfac = function DropdownDirective_Factory(t) { return new (t || DropdownDirective)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ElementRef)); };
DropdownDirective.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: DropdownDirective, selectors: [["", "clickOutside", ""]], hostBindings: function DropdownDirective_HostBindings(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵlistener("click", function DropdownDirective_click_HostBindingHandler($event) { return ctx.onClick($event.target); }, false, ɵngcc0.ɵɵresolveDocument);
    } }, inputs: { inputId: "inputId" }, outputs: { clickOutside: "clickOutside" } });
/** @nocollapse */
DropdownDirective.ctorParameters = () => [
    { type: ElementRef }
];
DropdownDirective.propDecorators = {
    clickOutside: [{ type: Output }],
    inputId: [{ type: Input }],
    onClick: [{ type: HostListener, args: ['document:click', ['$event.target'],] }]
};
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(DropdownDirective, [{
        type: Directive,
        args: [{
                selector: '[clickOutside]'
            }]
    }], function () { return [{ type: ɵngcc0.ElementRef }]; }, { clickOutside: [{
            type: Output
        }], 
    /**
     * @param {?} targetElement
     * @return {?}
     */
    onClick: [{
            type: HostListener,
            args: ['document:click', ['$event.target']]
        }], inputId: [{
            type: Input
        }] }); })();
if (false) {
    /** @type {?} */
    DropdownDirective.prototype.clickOutside;
    /** @type {?} */
    DropdownDirective.prototype.inputId;
    /**
     * @type {?}
     * @private
     */
    DropdownDirective.prototype._elementRef;
}

/**
 * @fileoverview added by tsickle
 * Generated from: lib/ngx-autocomplete.module.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
class NgxAutocompleteModule {
}
NgxAutocompleteModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: NgxAutocompleteModule });
NgxAutocompleteModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function NgxAutocompleteModule_Factory(t) { return new (t || NgxAutocompleteModule)(); }, imports: [[
            CommonModule
        ]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(NgxAutocompleteModule, { declarations: function () { return [NgxAutocompleteComponent, DropdownDirective]; }, imports: function () { return [CommonModule]; }, exports: function () { return [NgxAutocompleteComponent]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(NgxAutocompleteModule, [{
        type: NgModule,
        args: [{
                declarations: [
                    NgxAutocompleteComponent,
                    DropdownDirective
                ],
                imports: [
                    CommonModule
                ],
                exports: [
                    NgxAutocompleteComponent
                ]
            }]
    }], null, null); })();

/**
 * @fileoverview added by tsickle
 * Generated from: public-api.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * Generated from: ngx-angular-autocomplete.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */

export { NgxAutocompleteComponent, NgxAutocompleteModule, NgxAutocompleteService, DropdownDirective as ɵa };
